services/supervisor/src/transactions.rs

```diff
@@ -534,11 +718,130 @@ impl Supervisor {
         // Mark deploy as failed.
         schema
             .deploy_states
-            .put(&deploy_request, DeployState::Failed { height, error });
+            .put(&deploy_request, AsyncEventState::Failed { height, error });
 
         // Remove artifact from pending deployments: since we require
         // a confirmation from every node, failure for one node means failure
         // for the whole network.
         schema.pending_deployments.remove(&deploy_request.artifact);
     }
+
+    /// Confirms a local migration success by the given author's public key and checks
+    /// if all the confirmations are collected. If so, commits the migration.
+    /// If migration state hash differs from the expected one, migration fails though,
+    /// and `fail_migration` method is invoked.
+    fn confirm_migration(
+        &self,
+        mut context: ExecutionContext<'_>,
+        request: MigrationRequest,
+        state_hash: Hash,
+        author: PublicKey,
+    ) -> Result<(), ExecutionError> {
+        let core_schema = context.data().for_core();
+
+        let mut schema = SchemaImpl::new(context.service_data());
+
+        let mut state = schema
+            .migration_states
+            .get(&request)
+            .expect("BUG: Attempt to confirm a migration which does not have a stored state");
+
+        // Verify that state hash does match expected one.
+        if let Err(error) = state.add_state_hash(state_hash) {
+            // Hashes do not match, rollback the migration.
+            drop(schema); // Required for the context reborrow.
```

Comment:
Nit: Is this `drop` still required, or is it a refactoring artifact? As far as I see, the context is not reborrowed within *this* method.

Reply:
`self.fail_migration(context, request, error, initiate_rollback)` fails without `drop`.